test(designer): #1304 follow-up — ScreenItemsView autosave / draft / lock 系統合 test

## 背景

#1304 (closed by PR #1313) で ScreenItemsView の **render scaffold** (Step 1) を 6 test で確立、Step 2 (編集動作 test) は **#1314** で trace。本 ISSUE では Step 3 (autosave / draft / lock 系) を扱う。

#1304 本文で「Step 3 は mock 複雑、別 ISSUE で漸進対応」と明示されていたが、PR #1313 close 時点で起票漏れ。本 ISSUE で正式 trace 確立 (2026-05-24、memory `feedback_closed_issue_note_not_trace.md` 反映)。

## 課題

PR #1313 で確立した leaf-only mock 戦略 (hooks 実体 / mcpBridge / store / schema を vi.mock) では autosave / draft / lock 系の検証ができない。これらは hook (`useResourceEditor` / `useEditSession`) の内部状態遷移と mcpBridge の双方向通信に依存するため、追加 mock 戦略が必要。

## スコープ

| step | 内容 | 対象ファイル |
|---|---|---|
| 1 | autosave 動作 test (debounce / 連続編集 / 保存契約) | `ScreenItemsView.test.tsx` 拡張 or 新 file 分割 |
| 2 | edit-session draft 切替 test (resume / discard / 並行 draft) | 同上 |
| 3 | lock 系 test (sessionId 取得 / lock 喪失 / SaveConflictDialog 開く条件) | 同上 |
| 4 | mcpBridge mock を「単純 fixture 返却」から「状態遷移シミュレータ」に拡張 (or msw 等の library 導入検討) | mock 拡張 |

## スコープ外

- e2e (Playwright) 追加 — Vite crash 既知問題で別ライン
- backend connection の実 integration test — MCP E2E (frontend/e2e/mcp/) 既存 line で対応

## 依存

- 前提: #1304 (PR #1313) merge 済 (Step 1 scaffold 確立)
- 関連: #1314 (Step 2 編集動作 test) — 並行進行可能、test file 共有

## priority

`priority: low` (autosave / draft / lock は spec / e2e 層で部分カバー、本 ISSUE は component test 層での補強)

## 関連

- 親 (Step 1、closed): #1304 (PR #1313)
- 兄弟 (Step 2): #1314
- 参考 (edit-session-draft spec): `docs/spec/edit-session-draft.md` (#683 / #684)

## 起票経緯

PR #1313 (#1304) merge 後、ISSUE 本文 Step 3 の follow-up 起票漏れに気付き本 ISSUE 起票 (2026-05-24)。

memory `feedback_closed_issue_note_not_trace.md` 鉄則 0 違反防止のため、closed ISSUE 本文の「別 ISSUE で対応」記述は **必ず open ISSUE で trace 確立**する原則を再確認。

# 独立レビュー (Opus subagent、PR #1326)

## 総評

設計 (ISSUE #1315 comment) と実装の対応は誠実で、leaf-only mock 戦略堅持・既存 18 test の後方互換維持・dialog mock の marker 出力化・schemas/ 不変・OOM 抑止のいずれも要件を満たしている。test 22-25 (draft / saveConflict / sessionId sanity) は本質契約を正しく捉えており高品質。

ただし test 19/20 (autosave debounce) は **full-file 実行で 25% flake (20 回中 5 回失敗 — #19 ×1、#20 ×2、それ以外含む混合)** を実測した。isolation (`-t "autosave debounce"`) では 20/20 pass だが、Step 1/2 を先行実行する full-file では再現する。PR description の「25/25 pass」「duration 2.2s」報告は isolation 寄りの結果で、CI / regression 中で発火する不安定性を覆い隠している。これは memory `feedback_e2e_flake_isolation_vs_full_run.md` の anti-pattern にそのまま該当。**Must-fix 1 件**。

その他は Should-fix 1 件 (test 21 の root-cause 不明な「flush 部分発生許容」、本来 fake timer 切替タイミングで解消可能)、Nit 軽微 (typo 等)。

## Must-fix (merge ブロッカー)

### MF-1. test 19/20 の full-file 実行 flake (25% 失敗率)

**再現**:

- `npm run test:unit -- ScreenItemsView.test.tsx --run` を 20 回連続実行 → **15 pass / 5 fail**
  - 失敗ケース: `AssertionError: expected 2 to be 1` (test #19 or #20、`editSession.update` が debounce 圧縮されず 2 回発火)
- `npx vitest run -t "autosave debounce" ScreenItemsView.test.tsx` (isolation) → **20/20 pass**

**根本原因**:

`vi.useFakeTimers({ shouldAdvanceTime: true })` は **fake timer を有効にしつつ、setTimeout コールバックを real wall-clock の進行に合わせて発火させる** mode。Step 1/2 の先行 test で event loop が温まると以下の race が発生する:

```ts
await act(async () => {
  fireEvent.change(idInput!, { target: { value: "x1" } });   // ① debounce timer A set (real time t0)
});
// real time が host load で +50ms 進行することがある
await act(async () => {
  await vi.advanceTimersByTimeAsync(100);                     // ② fake timer +100ms 進めるが、real-time 経過分も加算される
});
// この時点で timer A が既に flush 済 (300ms 経過とみなされる) → editSession.update 1 回目発火
await act(async () => {
  fireEvent.change(idInput!, { target: { value: "x2" } });   // ③ debounce timer B set (timer A は既に flush 済なので reset されない)
});
await act(async () => {
  await vi.advanceTimersByTimeAsync(300);                     // ④ timer B 発火 → editSession.update 2 回目
});
// 結果: 2 回呼ばれて assertion 失敗
```

`shouldAdvanceTime: true` のドキュメント (vitest) も「実時間 + fake timer 二重計上の罠あり」と warn している。本 test の本質契約 (debounce が圧縮されて 1 回に集約される) と実装手段が矛盾している。

**修正案** (どちらか):

**(A) `shouldAdvanceTime: false` (デフォルト) に切替 + setTimeout を完全に fake 制御**

```ts
vi.useFakeTimers();  // shouldAdvanceTime 指定なし = false
// fireEvent.change / advanceTimersByTimeAsync 順序を厳密に管理
// waitFor が polling できないので act() + manual flush で構成
```

trade-off: waitFor が動かないので render flush は `await act(async () => { /* no-op */ })` + microtask 強制で代替する必要あり。

**(B) fake timer を使わず、debounce duration (300ms) を待つ実時間 wait に置換**

```ts
await new Promise((resolve) => setTimeout(resolve, 350));
const updateCalls = vi.mocked(mcpBridge.request).mock.calls.filter(([m]) => m === "editSession.update");
expect(updateCalls.length).toBe(1);
```

trade-off: test #19/#20 が +350ms ずつ遅延 (全体 +0.7s 程度)、flake 解消は確実。Step 2 既存 test がそうしているように、本 component の autosave test 規模なら real timer + 短 wait が安全。

**(C) test #19/#20 を削除** — ただし design 観点が消えるので非推奨

**推奨**: (B)。実装コストが最小で flake が完全消滅、duration impact も許容範囲。

flake 25% は merge 後 CI / regression chain で混乱を量産し、bisect 時に誤導につながる。Must-fix として merge ブロック。

## Should-fix (本 PR スコープ内、検討推奨)

### SF-1. test #21 の「flush 発生有無を許容」緩和は root-cause を回避している

`ScreenItemsView.tsx:323-338` の `handleSave` は明確に:

1. `clearTimeout(draftUpdateTimer.current)` で pending を cancel
2. `await mcpBridge.request("editSession.update", ...)` で同期 flush 1 回
3. `await actions.save()` で `editSession.save` 1 回
4. `if (conflicted || failed) return; await postSave()`

つまり保存 click 後の `editSession.update` 回数は **厳密に 1**。PR description の「shouldAdvanceTime により debounce が部分的に flush される場合がある」は MF-1 と同じ fake-timer 起因の現象で、本来 `handleSave` 内 clearTimeout が機能している限り pending debounce flush は発生しない。

MF-1 を (B) real timer 化で解消すれば、test #21 も以下に厳格化できる:

```ts
expect(afterSaveUpdateCount).toBe(1);  // >= 1 ではなく == 1
```

このまま緩和を残すと、将来 `handleSave` の clearTimeout が消えた regression を test #21 が検出できなくなる (zombie update が 2 回でも pass する)。

## Nit (任意)

### N-1. test 19-25 コメントブロックの URL typo

`ScreenItemsView.test.tsx:810`:

```
//   https://github.com/csilost2001/issues/1315#issuecomment-4528558182
```

正しくは `https://github.com/csilost2001/harmony/issues/1315#issuecomment-4528558182` (repo 部分 `/harmony` が抜けている)。設計参照リンクが壊れている。

### N-2. `makeDefaultBridgeImpl` の overrides 型は narrow すぎ

```ts
overrides: Partial<Record<string, (params: unknown) => unknown>> = {}
```

`Promise.resolve(overrides[method]!(params))` で wrap しているため、override 内で Promise を返したい (例: reject simulation) ケースが書けない。将来 #1315 系の lock 喪失 test を追加する時に拡張余地が無くなる。signature を:

```ts
overrides: Partial<Record<string, (params: unknown) => unknown | Promise<unknown>>> = {}
```

にして、`if (overrides[method]) return Promise.resolve(overrides[method]!(params))` のままで Promise 入れ子も受ける形にすれば、後方互換で拡張可。

### N-3. test #25 の sanity 観点は冗長

`mcpBridge.getSessionId` の sanity check ということになっているが、実際の assertion は `editSession.list` の params 形だけで、`getSessionId` の戻り値検証は無い。設計 comment の意図 (sessionId が hooks に渡るかの sanity) を満たすには `vi.mocked(mcpBridge.getSessionId).mock.calls.length > 0` も assert すべき。

ただし本 PR scope では既存 18 test と同様の params 形 sanity で十分機能している。テスト価値は薄いが merge ブロックではない。

## スコープ外指摘 (別 ISSUE 候補)

なし。MF-1 / SF-1 / N-1〜N-3 はすべて本 PR 内で吸収可能 (鉄則 1)。

## 検証 (本 review 実施分)

| 項目 | 結果 |
|------|------|
| `gh pr diff 1326 --name-only` | 1 ファイル (`frontend/src/components/screen-items/ScreenItemsView.test.tsx`) |
| `git diff origin/main..HEAD -- schemas/` | empty (schema governance ✅) |
| `npm run test:unit -- ScreenItemsView.test.tsx --run` × 1 回 | 1 failed (#20) |
| 同上 × 10 回連続 | 7 pass / 3 fail (#19 ×1, #20 ×2) |
| 同上 × 20 回連続 | 15 pass / 5 fail (25% flake) |
| `npx vitest run -t "autosave debounce" ScreenItemsView.test.tsx` × 20 回 | 20 pass / 0 fail (isolation OK) |
| `npm run test:unit -- src/components/screen-items src/components/editing --run` (82 test regression) | 1 failed (#20、同根) |
| `git log origin/main..HEAD --oneline` | 1 commit (`ae883928`) |

## マージ判定

❌ reject (MF-1 解消後 conditional approve)

MF-1 (autosave debounce flake) は merge 後 CI / 他 PR の regression 判定を 25% 確率で汚すため必ず修正必要。SF-1 と N-1〜N-3 は MF-1 修正と同時に対応推奨 (どれも 1-2 行 / 同 file 内で完結、別 ISSUE 化は鉄則 0 違反)。

---

_Generated by [Claude Code] (Opus subagent independent review, 2026-05-24)_

--- a/frontend/src/components/screen-items/ScreenItemsView.test.tsx
+++ b/frontend/src/components/screen-items/ScreenItemsView.test.tsx
@@ -807,7 +807,7 @@ describe("ScreenItemsView (Step 2: 編集動作)", () => {
 // tests (Step 3: autosave / draft / lock — #1315)
 //
 // 設計: ISSUE #1315 comment
-//   https://github.com/csilost2001/issues/1315#issuecomment-4528558182
+//   https://github.com/csilost2001/harmony/issues/1315#issuecomment-4528558182
 //
 // 観点 (test 19-25):
 //   - autosave debounce (#19-#21): updateWithDraft 経由の editSession.update が
@@ -826,7 +826,7 @@ describe("ScreenItemsView (Step 2: 編集動作)", () => {
  * デフォルト挙動 (それぞれ妥当な response) で返す factory。
  * 個別 test は必要に応じて mockImplementation を差し替えて override する。
  */
-function makeDefaultBridgeImpl(overrides: Partial<Record<string, (params: unknown) => unknown>> = {}) {
+function makeDefaultBridgeImpl(overrides: Partial<Record<string, (params: unknown) => unknown | Promise<unknown>>> = {}) {
   return (method: string, params?: unknown) => {
     if (overrides[method]) {
       return Promise.resolve(overrides[method]!(params));
@@ -858,16 +858,20 @@ function makeDefaultBridgeImpl(overrides: Partial<Record<string, (params: unknow
 }
 
 describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
-  beforeEach(() => {
+  beforeEach(async () => {
     // editing mode サポート (Step 2 と同等の base impl + editSession.save/update を追加)
     vi.mocked(mcpBridge.request).mockImplementation(makeDefaultBridgeImpl() as never);
-  });
 
-  afterEach(() => {
-    // fake timer を有効にした test の後、real timer に戻す (test 内でも cleanup 済みの場合は no-op)
-    if (vi.isFakeTimers()) {
-      vi.useRealTimers();
-    }
+    // Step 1/2 test が localStorage に draft を残す + setTimeout(300) が pending のまま
+    // 終わるため、Step 3 の autosave debounce 観測前に必ず flush + clear する。
+    //   - localStorage.clear(): 旧 draft (effects 等) が useResourceEditor.reload で
+    //     restoreされてしまい計測前の payload を汚染するのを防ぐ
+    //   - 350ms 待機: 前 test の updateSilentWithDraft 由来の setTimeout(300) が
+    //     unmount 後も pending で残っており、本 test の計測窓に紛れ込むのを防ぐ
+    //   (PR #1326 review MF-1: fake timer → real timer 切替に伴う追加防御)
+    localStorage.clear();
+    await new Promise((resolve) => setTimeout(resolve, 350));
+    vi.mocked(mcpBridge.request).mockClear();
   });
 
   // -------------------------------------------------------------------------
@@ -890,16 +894,15 @@ describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
       );
       expect(idInput).not.toBeNull();
 
-      // fake timer 切替 (shouldAdvanceTime: true で waitFor の polling も動作可)
-      vi.useFakeTimers({ shouldAdvanceTime: true });
-
       await act(async () => {
         fireEvent.change(idInput!, { target: { value: "customerCode" } });
       });
 
-      // debounce 完了 (300ms) を超えて 350ms 進める
+      // debounce 完了 (300ms) を超えて 350ms 実時間待機
+      // (fake timer + shouldAdvanceTime は full-file 実行で real-time 二重計上による
+      //  事前 flush flake を起こすため real timer に統一、PR #1326 review MF-1)
       await act(async () => {
-        await vi.advanceTimersByTimeAsync(350);
+        await new Promise((resolve) => setTimeout(resolve, 350));
       });
 
       // editSession.update が **1 回だけ** 呼ばれたこと
@@ -923,16 +926,14 @@ describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
       );
       expect(idInput).not.toBeNull();
 
-      vi.useFakeTimers({ shouldAdvanceTime: true });
-
       // 1 回目編集
       await act(async () => {
         fireEvent.change(idInput!, { target: { value: "x1" } });
       });
 
-      // 100ms 経過 (debounce 未 flush)
+      // 100ms 経過 (debounce 未 flush) — real timer wait
       await act(async () => {
-        await vi.advanceTimersByTimeAsync(100);
+        await new Promise((resolve) => setTimeout(resolve, 100));
       });
 
       // 2 回目編集 (previous timer reset)
@@ -940,9 +941,9 @@ describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
         fireEvent.change(idInput!, { target: { value: "x2" } });
       });
 
-      // 300ms 進める (合計 400ms だが 2 回目から 300ms = flush)
+      // 300ms 進める (合計 400ms だが 2 回目から 300ms = flush) — real timer wait
       await act(async () => {
-        await vi.advanceTimersByTimeAsync(300);
+        await new Promise((resolve) => setTimeout(resolve, 300));
       });
 
       // editSession.update は 1 回だけ呼ばれること (debounce 圧縮)
@@ -966,19 +967,16 @@ describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
       );
       expect(idInput).not.toBeNull();
 
-      vi.useFakeTimers({ shouldAdvanceTime: true });
-
       // 編集 → debounce timer start
       await act(async () => {
         fireEvent.change(idInput!, { target: { value: "x" } });
       });
 
-      // 100ms だけ進める (実時間環境では shouldAdvanceTime により debounce が
-      // 部分的に flush される場合があるが、本 test の本質的契約は
-      // 「保存後に editSession.save が呼ばれ、editSession.update が flush され、
-      //  保存後に新たな debounce 発火がない」点なので flush 発生有無は許容する)
+      // 100ms だけ実時間 wait (debounce 300ms 未満なので pending のまま)
+      // 続いて 即座に「保存」click → handleSave 内で clearTimeout + 直接 flush するため
+      // editSession.update は **厳密に 1 回** だけ呼ばれる
       await act(async () => {
-        await vi.advanceTimersByTimeAsync(100);
+        await new Promise((resolve) => setTimeout(resolve, 100));
       });
 
       // 「保存」button click
@@ -996,17 +994,17 @@ describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
         expect(saveCalls.length).toBe(1);
       }, { timeout: 3000 });
 
-      // editSession.update も少なくとも 1 回は flush されている
-      //   (handleSave 内で同期 await + 場合により先行 debounce flush)
+      // editSession.update は handleSave 内で clearTimeout + 直接 flush するため
+      // **厳密に 1 回** (PR #1326 review SF-1: MF-1 解消で flake 要因消滅、tighten 可能)
       const afterSaveUpdateCount = vi.mocked(mcpBridge.request).mock.calls.filter(
         ([m]) => m === "editSession.update"
       ).length;
-      expect(afterSaveUpdateCount).toBeGreaterThanOrEqual(1);
+      expect(afterSaveUpdateCount).toBe(1);
 
-      // さらに 400ms 進めても update は増えない (handleSave 内で timer cleared)
+      // さらに 400ms 実時間待っても update は増えない (handleSave 内で timer cleared)
       // = 保存処理の本質契約: pending debounce が save 後にゾンビ発火しない
       await act(async () => {
-        await vi.advanceTimersByTimeAsync(400);
+        await new Promise((resolve) => setTimeout(resolve, 400));
       });
       const finalUpdateCount = vi.mocked(mcpBridge.request).mock.calls.filter(
         ([m]) => m === "editSession.update"

--- a/frontend/src/components/screen-items/ScreenItemsView.test.tsx
+++ b/frontend/src/components/screen-items/ScreenItemsView.test.tsx
@@ -125,11 +125,19 @@ vi.mock("../../schemas/conventionsValidator", () => ({
 // ---------------------------------------------------------------------------
 // 重い子 component mocks (modal 系は DOM に出てこない方が test 軽い)
 // ---------------------------------------------------------------------------
+// Step 3: dialog mock を marker 出力型に変更 (props 駆動で testid 出力)
+// Step 1/2 の test は dialog 条件 (saveConflict truthy or showResumeDialog=true) を
+// 満たさないため marker は出現せず、後方互換性を維持する。
 vi.mock("../editing/SaveConflictDialog", () => ({
-  SaveConflictDialog: () => null,
+  SaveConflictDialog: ({ conflict }: { conflict: unknown }) =>
+    conflict ? <div data-testid="save-conflict-dialog" /> : null,
 }));
 vi.mock("../editing/ResumeOrDiscardDialog", () => ({
-  ResumeOrDiscardDialog: () => null,
+  ResumeOrDiscardDialog: ({ onResume }: { onResume?: () => void }) => (
+    <div data-testid="resume-or-discard-dialog">
+      <button data-testid="resume-or-discard-resume" onClick={onResume}>resume</button>
+    </div>
+  ),
 }));
 vi.mock("./ScreenItemCandidatesModal", () => ({
   ScreenItemCandidatesModal: () => null,
@@ -791,6 +799,354 @@ describe("ScreenItemsView (Step 2: 編集動作)", () => {
         const rows = container.querySelectorAll("[aria-label*='を選択']:not([aria-label='全選択'])");
         expect(rows.length).toBe(2);
       }, { timeout: 3000 });
+    });
+  });
+});
+
+// ---------------------------------------------------------------------------
+// tests (Step 3: autosave / draft / lock — #1315)
+//
+// 設計: ISSUE #1315 comment
+//   https://github.com/csilost2001/issues/1315#issuecomment-4528558182
+//
+// 観点 (test 19-25):
+//   - autosave debounce (#19-#21): updateWithDraft 経由の editSession.update が
+//     300ms debounce で 1 回に圧縮されること、保存時には即 flush されること
+//   - edit-session draft (#22-#23): readonly mode で自分の Active session が
+//     残っていた場合に ResumeOrDiscardDialog が出現、Resume click で startEditing
+//     (= editSession.create) が呼ばれること
+//   - lock / saveConflict (#24-#25): backend が {ok:false, conflict:...} を返した
+//     場合に SaveConflictDialog が表示されること、editSession.list が正しい
+//     {resourceType, resourceId} で呼ばれていること (sanity)
+// ---------------------------------------------------------------------------
+
+/**
+ * editing mode 用の mock implementation を返す。
+ * editSession.list / editSession.create / editSession.save / editSession.update を
+ * デフォルト挙動 (それぞれ妥当な response) で返す factory。
+ * 個別 test は必要に応じて mockImplementation を差し替えて override する。
+ */
+function makeDefaultBridgeImpl(overrides: Partial<Record<string, (params: unknown) => unknown>> = {}) {
+  return (method: string, params?: unknown) => {
+    if (overrides[method]) {
+      return Promise.resolve(overrides[method]!(params));
+    }
+    if (method === "editSession.list") {
+      return Promise.resolve({ sessions: [] });
+    }
+    if (method === "editSession.create") {
+      return Promise.resolve({
+        editSession: {
+          id: "mock-session-id",
+          resourceType: "screen-item",
+          resourceId: "test-screen-id",
+          state: "Active",
+          participants: { "test-session-id": { role: "Edit" } },
+          sequence: 0,
+          payload: null,
+        },
+      });
+    }
+    if (method === "editSession.save") {
+      return Promise.resolve({ ok: true });
+    }
+    if (method === "editSession.update") {
+      return Promise.resolve(null);
+    }
+    return Promise.resolve(null);
+  };
+}
+
+describe("ScreenItemsView (Step 3: autosave / draft / lock)", () => {
+  beforeEach(() => {
+    // editing mode サポート (Step 2 と同等の base impl + editSession.save/update を追加)
+    vi.mocked(mcpBridge.request).mockImplementation(makeDefaultBridgeImpl() as never);
+  });
+
+  afterEach(() => {
+    // fake timer を有効にした test の後、real timer に戻す (test 内でも cleanup 済みの場合は no-op)
+    if (vi.isFakeTimers()) {
+      vi.useRealTimers();
+    }
+  });
+
+  // -------------------------------------------------------------------------
+  describe("autosave debounce (#19-#21)", () => {
+    it("19. 編集 1 回 → 300ms 後に editSession.update が 1 回呼ばれる", async () => {
+      const { loadScreenItems } = await import("../../store/screenItemsStore");
+      vi.mocked(loadScreenItems).mockResolvedValueOnce(FIXTURE_SCREEN_ITEMS_ONE_ITEM);
+
+      const { container } = renderWithRouter();
+      await startEditingMode(container);
+
+      // editing mode 移行時に editSession.create が呼ばれているので clear して計測開始
+      vi.mocked(mcpBridge.request).mockClear();
+
+      // ID 列の input (Identifier 入力欄) を 1 回編集 → updateSilentWithDraft 経由で
+      // debounce timer が start する
+      // tbody の最初の行の最初の text input (= ID 入力欄) を取得
+      const idInput = container.querySelector<HTMLInputElement>(
+        "tbody tr input[type='text']"
+      );
+      expect(idInput).not.toBeNull();
+
+      // fake timer 切替 (shouldAdvanceTime: true で waitFor の polling も動作可)
+      vi.useFakeTimers({ shouldAdvanceTime: true });
+
+      await act(async () => {
+        fireEvent.change(idInput!, { target: { value: "customerCode" } });
+      });
+
+      // debounce 完了 (300ms) を超えて 350ms 進める
+      await act(async () => {
+        await vi.advanceTimersByTimeAsync(350);
+      });
+
+      // editSession.update が **1 回だけ** 呼ばれたこと
+      const updateCalls = vi.mocked(mcpBridge.request).mock.calls.filter(
+        ([m]) => m === "editSession.update"
+      );
+      expect(updateCalls.length).toBe(1);
+    });
+
+    it("20. 連続 2 編集 (各 100ms 間隔) → debounce で 1 回に圧縮される", async () => {
+      const { loadScreenItems } = await import("../../store/screenItemsStore");
+      vi.mocked(loadScreenItems).mockResolvedValueOnce(FIXTURE_SCREEN_ITEMS_ONE_ITEM);
+
+      const { container } = renderWithRouter();
+      await startEditingMode(container);
+
+      vi.mocked(mcpBridge.request).mockClear();
+
+      const idInput = container.querySelector<HTMLInputElement>(
+        "tbody tr input[type='text']"
+      );
+      expect(idInput).not.toBeNull();
+
+      vi.useFakeTimers({ shouldAdvanceTime: true });
+
+      // 1 回目編集
+      await act(async () => {
+        fireEvent.change(idInput!, { target: { value: "x1" } });
+      });
+
+      // 100ms 経過 (debounce 未 flush)
+      await act(async () => {
+        await vi.advanceTimersByTimeAsync(100);
+      });
+
+      // 2 回目編集 (previous timer reset)
+      await act(async () => {
+        fireEvent.change(idInput!, { target: { value: "x2" } });
+      });
+
+      // 300ms 進める (合計 400ms だが 2 回目から 300ms = flush)
+      await act(async () => {
+        await vi.advanceTimersByTimeAsync(300);
+      });
+
+      // editSession.update は 1 回だけ呼ばれること (debounce 圧縮)
+      const updateCalls = vi.mocked(mcpBridge.request).mock.calls.filter(
+        ([m]) => m === "editSession.update"
+      );
+      expect(updateCalls.length).toBe(1);
+    });
+
+    it("21. 「保存」click → editSession.save 呼ばれる + editSession.update flush + 保存後 timer 残存しない", async () => {
+      const { loadScreenItems } = await import("../../store/screenItemsStore");
+      vi.mocked(loadScreenItems).mockResolvedValueOnce(FIXTURE_SCREEN_ITEMS_ONE_ITEM);
+
+      const { container } = renderWithRouter();
+      await startEditingMode(container);
+
+      vi.mocked(mcpBridge.request).mockClear();
+
+      const idInput = container.querySelector<HTMLInputElement>(
+        "tbody tr input[type='text']"
+      );
+      expect(idInput).not.toBeNull();
+
+      vi.useFakeTimers({ shouldAdvanceTime: true });
+
+      // 編集 → debounce timer start
+      await act(async () => {
+        fireEvent.change(idInput!, { target: { value: "x" } });
+      });
+
+      // 100ms だけ進める (実時間環境では shouldAdvanceTime により debounce が
+      // 部分的に flush される場合があるが、本 test の本質的契約は
+      // 「保存後に editSession.save が呼ばれ、editSession.update が flush され、
+      //  保存後に新たな debounce 発火がない」点なので flush 発生有無は許容する)
+      await act(async () => {
+        await vi.advanceTimersByTimeAsync(100);
+      });
+
+      // 「保存」button click
+      const saveBtn = container.querySelector<HTMLButtonElement>('[data-testid="edit-mode-save"]');
+      expect(saveBtn).not.toBeNull();
+      await act(async () => {
+        fireEvent.click(saveBtn!);
+      });
+
+      // editSession.save が呼ばれること
+      await waitFor(() => {
+        const saveCalls = vi.mocked(mcpBridge.request).mock.calls.filter(
+          ([m]) => m === "editSession.save"
+        );
+        expect(saveCalls.length).toBe(1);
+      }, { timeout: 3000 });
+
+      // editSession.update も少なくとも 1 回は flush されている
+      //   (handleSave 内で同期 await + 場合により先行 debounce flush)
+      const afterSaveUpdateCount = vi.mocked(mcpBridge.request).mock.calls.filter(
+        ([m]) => m === "editSession.update"
+      ).length;
+      expect(afterSaveUpdateCount).toBeGreaterThanOrEqual(1);
+
+      // さらに 400ms 進めても update は増えない (handleSave 内で timer cleared)
+      // = 保存処理の本質契約: pending debounce が save 後にゾンビ発火しない
+      await act(async () => {
+        await vi.advanceTimersByTimeAsync(400);
+      });
+      const finalUpdateCount = vi.mocked(mcpBridge.request).mock.calls.filter(
+        ([m]) => m === "editSession.update"
+      ).length;
+      expect(finalUpdateCount).toBe(afterSaveUpdateCount);
+    });
+  });
+
+  // -------------------------------------------------------------------------
+  describe("edit-session draft (#22-#23)", () => {
+    it("22. editSession.list が my Active session を返す → ResumeOrDiscardDialog marker 表示", async () => {
+      // editSession.list を「my Active session 1 件」を返すよう差し替え
+      vi.mocked(mcpBridge.request).mockImplementation(
+        makeDefaultBridgeImpl({
+          "editSession.list": () => ({
+            sessions: [
+              {
+                id: "es-existing",
+                resourceType: "screen-item",
+                resourceId: "test-screen-id",
+                state: "Active",
+                participants: { "test-session-id": { role: "Edit" } },
+                sequence: 0,
+                payload: null,
+              },
+            ],
+          }),
+        }) as never,
+      );
+
+      const { container } = renderWithRouter();
+
+      // readonly mode のままで Resume dialog 表示を待つ (startEditingMode は呼ばない)
+      await waitFor(() => {
+        const marker = container.querySelector('[data-testid="resume-or-discard-dialog"]');
+        expect(marker).not.toBeNull();
+      }, { timeout: 3000 });
+    });
+
+    it("23. Resume click → startEditing (editSession.create) が呼ばれる", async () => {
+      vi.mocked(mcpBridge.request).mockImplementation(
+        makeDefaultBridgeImpl({
+          "editSession.list": () => ({
+            sessions: [
+              {
+                id: "es-existing",
+                resourceType: "screen-item",
+                resourceId: "test-screen-id",
+                state: "Active",
+                participants: { "test-session-id": { role: "Edit" } },
+                sequence: 0,
+                payload: null,
+              },
+            ],
+          }),
+        }) as never,
+      );
+
+      const { container } = renderWithRouter();
+
+      await waitFor(() => {
+        const marker = container.querySelector('[data-testid="resume-or-discard-dialog"]');
+        expect(marker).not.toBeNull();
+      }, { timeout: 3000 });
+
+      // resume click 前の create 呼び出し履歴をクリア
+      vi.mocked(mcpBridge.request).mockClear();
+
+      const resumeBtn = container.querySelector<HTMLButtonElement>(
+        '[data-testid="resume-or-discard-resume"]'
+      );
+      expect(resumeBtn).not.toBeNull();
+      await act(async () => {
+        fireEvent.click(resumeBtn!);
+      });
+
+      await waitFor(() => {
+        const createCalls = vi.mocked(mcpBridge.request).mock.calls.filter(
+          ([m]) => m === "editSession.create"
+        );
+        expect(createCalls.length).toBeGreaterThanOrEqual(1);
+      }, { timeout: 3000 });
+    });
+  });
+
+  // -------------------------------------------------------------------------
+  describe("lock / saveConflict (#24-#25)", () => {
+    it("24. 保存 click → backend 返却 {ok:false, conflict} → SaveConflictDialog marker 表示", async () => {
+      vi.mocked(mcpBridge.request).mockImplementation(
+        makeDefaultBridgeImpl({
+          "editSession.save": () => ({
+            ok: false,
+            conflict: {
+              other: {
+                savedAt: "2026-01-02T00:00:00Z",
+                ownerSessionId: "other-session",
+              },
+            },
+          }),
+        }) as never,
+      );
+
+      const { loadScreenItems } = await import("../../store/screenItemsStore");
+      vi.mocked(loadScreenItems).mockResolvedValueOnce(FIXTURE_SCREEN_ITEMS_ONE_ITEM);
+
+      const { container } = renderWithRouter();
+      await startEditingMode(container);
+
+      // 「保存」click
+      const saveBtn = container.querySelector<HTMLButtonElement>('[data-testid="edit-mode-save"]');
+      expect(saveBtn).not.toBeNull();
+      await act(async () => {
+        fireEvent.click(saveBtn!);
+      });
+
+      // SaveConflictDialog marker が表示されること
+      await waitFor(() => {
+        const marker = container.querySelector('[data-testid="save-conflict-dialog"]');
+        expect(marker).not.toBeNull();
+      }, { timeout: 3000 });
+    });
+
+    it("25. mcpBridge.request の calls 履歴に editSession.list {resourceType:'screen-item', resourceId:'test-screen-id'} が含まれる (sanity)", async () => {
+      renderWithRouter();
+
+      // ResumeOrDiscardDialog effect が editSession.list を呼ぶことを待つ
+      await waitFor(() => {
+        const listCalls = vi.mocked(mcpBridge.request).mock.calls.filter(
+          ([m]) => m === "editSession.list"
+        );
+        expect(listCalls.length).toBeGreaterThanOrEqual(1);
+      }, { timeout: 3000 });
+
+      // params が想定通り {resourceType, resourceId} を持つ call が存在すること
+      const matchedCall = vi.mocked(mcpBridge.request).mock.calls.find(([m, params]) => {
+        if (m !== "editSession.list") return false;
+        const p = params as { resourceType?: string; resourceId?: string } | undefined;
+        return p?.resourceType === "screen-item" && p?.resourceId === "test-screen-id";
+      });
+      expect(matchedCall).toBeDefined();
     });
   });
 });

